### ISSUE TITLE
confusion in code using a mix of abilities

### ISSUE THREAD

**Issue by stew** [unisonweb/unison#2344]
Tried to minimize this as much as possible. I'm suprised that `nate` needs to be a tuple, but it I can't seem to eliminate that. In the following code, `grumpy` does compile, but `sneezy` and `doc` both complain about the ultimate term in the expression

```
unique ability Nate where
  nate: (Boolean, Nat)
  antiNate: ()

grumpy: (Nat -> {d}()) -> '{d, Nate}()
grumpy dee _ =
  antiNate
  dee 1
  (_, _) = nate
  ()

sneezy: (Nat -> {d}()) -> '{d, Nate}()
sneezy dee _ =
  (_, _) = nate
  antiNate
  dee 1

doc: (Nat -> {d}()) -> '{d, Nate}()
doc dee _ =
  (_, _) = nate
  dee 1
  antiNate
```

The error for `sneezy`:
```
  The expression in red needs the {d36} ability, but this location does not have access to any abilities.
  
     17 |   dee 1
```

The error for `doc`:
```
  The expression in red needs the {Nate} ability, but this location does not have access to any abilities.
  
     17 |   antiNate

```

**Comment by dolio** [comment]
I suspect the importance of the tuple is that it's pattern matching, not just a let. So these examples are actually more different than they first appear.

**Comment by stew** [comment]
oh, yes, if I avoid the pattern match and destructure the tupple with `at1`,`at2` then it works (so I'm unblocked :) )

**Comment by dolio** [comment]
Here's what seems to be going on. The destructuring bind desugars like this:

```
sneezy dee _ =
  (_, _) = nate
  antiNate
  dee 1

==>

let rec
  sneezy dee _ = match nate with
    (_, _) -> let rec
      _ = antiNate
      dee 1
```

Then, recursive lets aren't allowed to have effects, because the order of evaluation isn't well defined. However, this shouldn't actually be recursive.

**Comment by pchiusano** [comment]
@dolio ah

There's actually some logic in the typechecker that minimizes cycles - see the call to `minimize'` and the module `Components`. So the parser generates let rec blocks that match what the user typed, and then the typechecker handles breaking that up into minimal components with `let` used in place of `let rec` where possible.

However, I bet this logic just isn't prepared to deal with destructuring bind, and there might be a simple fix to `Components.minimize` that does the trick.

**Comment by dolio** [comment]
Doesn't the `minimize'` thing get called with the desugared code? It seems like the desugaring happens immediately in the parser.

The problem is that this shouldn't be a cycle at all, not that the cycle isn't minimal. It seems like the destructuring bind is forcing the block to be a cycle. But it should probably be the opposite; that destructuring binds should _not_ be cyclic.

**Comment by pchiusano** [comment]
Hmm, it looks like it's done in both places, in the parser and in the typechecker.

The logic in `minimize` will convert `let rec` to `let` where possible, so I think that logic just needs updating to deal with destructuring bind. 

You could also try to avoid building a needless `LetRec` in the first place in the parser, but even if that were done, I think you'd want to update `minimize` to handle this better - so that code written before the parser desugaring improvement can still typecheck (like during patch propagation).

**Comment by stew** [comment]
I'm running into this one again, here's the best I could minimize this example, which is weird in itself

```haskell
segal :  TVar (Map Text Text) -> {.base.IO} ()
segal state =
  mod: '{STM}()
  mod  _ = TVar.modify state (tasks -> Map.insert tid "asdf" tasks)

  tv = TVar.newIO ()
  tid = "asdf"
  ()

```

error:

```
  The expression in red needs the {base.IO} ability, but this location does not have access to any abilities.
  
      7 |   tv = TVar.newIO ()

```
(with `TVar.newIO ()` being the thing in red)

**Comment by pchiusano** [comment]
Okay, you are getting a terrible error message, but this variation typechecks, and I'll explain why:

```Haskell
segal :  TVar (Map Text Text) ->{IO} ()
segal state =
  mod: '{STM}()
  mod  _ = TVar.modify state (tasks -> Map.insert tid "asdf" tasks)
  tid = "asdf"
  tv = TVar.newIO ()
  ()
```

All I did was swap the order of the two lines declaring `tid` and `tv`.

What's going on? The function passed to `TVar.modify` references `tid`, which is declared later on in the function. The typechecker considers all bindings in between the `modify` and the forward reference to be part of a cycle, so that's `mod`, `tid`, `tv`. Cycles don't have access to any abilities, since order of effect evaluation becomes unclear. Here it might be fine since the forward reference is guarded by a lambda but the typechecker isn't that smart.

**Comment by pchiusano** [comment]
TL;DR: Forward-references create cycles of bindings, and cycles don't have access to any abilities. Workaround is to rearrange your bindings so as not to use forward references or limit how far forward your references are, so they don't capture other unrelated stuff in the cycle.

**Comment by pchiusano** [comment]
Going to close in favor of https://github.com/unisonweb/unison/issues/2407
